File: sdk/node/examples/test_apis.ts
/**
 * Test script to verify Orchestrator monitoring APIs are working correctly.
 * This test uses placeholder execution (no real LLM calls).
 */

import { Orchestrator, SubAgentConfig } from '@a3s-lab/code';

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function testOrchestratorAPIs(): Promise<boolean> {
  console.log('Testing Orchestrator Monitoring APIs...\n');

  // Test 1: Create Orchestrator
  console.log('1. Testing Orchestrator.create()...');
  let orch: Orchestrator;
  try {
    orch = Orchestrator.create();
    console.log('   ✓ Orchestrator created');
  } catch (e) {
    console.log(`   ✗ Failed: ${e}`);
    return false;
  }

  // Test 2: Create SubAgentConfig
  console.log('\n2. Testing SubAgentConfig...');
  let config: SubAgentConfig;
  try {
    config = {
      agentType: 'test',
      description: 'Test SubAgent',
      prompt: 'Test prompt',
      permissive: true,
      maxSteps: 3,
    };
    console.log('   ✓ SubAgentConfig created');
  } catch (e) {
    console.log(`   ✗ Failed: ${e}`);
    return false;
  }

  // Test 3: Spawn SubAgent
  console.log('\n3. Testing spawnSubagent()...');
  let handle;
  try {
    handle = orch.spawnSubagent(config);
    console.log(`   ✓ SubAgent spawned: ${handle.id}`);
  } catch (e) {
    console.log(`   ✗ Failed: ${e}`);
    return false;
  }

  // Test 4: Active count
  console.log('\n4. Testing activeCount()...');
  try {
    const count = orch.activeCount();
    console.log(`   ✓ Active count: ${count}`);
  } catch (e) {
    console.log(`   ✗ Failed: ${e}`);
    return false;
  }

  // Test 5: List SubAgents
  console.log('\n5. Testing listSubagents()...');
  try {
    const subagents = orch.listSubagents();
    console.log(`   ✓ Found ${subagents.length} SubAgent(s)`);
    if (subagents.length > 0) {
      const info = subagents[0];
      console.log(`      - ID: ${info.id}`);
      console.log(`      - Type: ${info.agentType}`);
      console.log(`      - State: ${info.state}`);
      console.log(`      - Description: ${info.description}`);
      if (info.currentActivity) {
        console.log(`      - Activity: ${info.currentActivity.activityType}`);
      }
    }
  } catch (e) {
    console.log(`   ✗ Failed: ${e}`);
    return false;
  }

  // Test 6: Get SubAgent info
  console.log('\n6. Testing getSubagentInfo()...');
  try {
    const info = orch.getSubagentInfo(handle.id);
    if (info) {
      console.log(`   ✓ Got info for ${info.id}`);
      console.log(`      - State: ${info.state}`);
      console.log(`      - Created: ${info.createdAt}`);
      console.log(`      - Updated: ${info.updatedAt}`);
    } else {
      console.log('   ✗ No info returned');
      return false;
    }
  } catch (e) {
    console.log(`   ✗ Failed: ${e}`);
    return false;
  }

  // Test 7: Get active activities
  console.log('\n7. Testing getActiveActivities()...');
  try {
    const activities = orch.getActiveActivities();
    console.log(`   ✓ Found ${activities.length} active activit(ies)`);
    for (const entry of activities) {
      console.log(`      - ${entry.id}: ${entry.activity.activityType}`);
    }
  } catch (e) {
    console.log(`   ✗ Failed: ${e}`);
    return false;
  }

  // Test 8: Get all states
  console.log('\n8. Testing getAllStates()...');
  try {
    const states = orch.getAllStates();
    console.log(`   ✓ Found ${states.length} state(s)`);
    for (const entry of states) {
      console.log(`      - ${entry.id}: ${entry.state}`);
    }
  } catch (e) {
    console.log(`   ✗ Failed: ${e}`);
    return false;
  }

  // Test 9: Pause SubAgent
  console.log('\n9. Testing pauseSubagent()...');
  try {
    orch.pauseSubagent(handle.id);
    await sleep(200);
    const info = orch.getSubagentInfo(handle.id);
    console.log(`   ✓ Paused, state: ${info?.state}`);
  } catch (e) {
    console.log(`   ✗ Failed: ${e}`);
    return false;
  }

  // Test 10: Resume SubAgent
  console.log('\n10. Testing resumeSubagent()...');
  try {
    orch.resumeSubagent(handle.id);
    await sleep(200);
    const info = orch.getSubagentInfo(handle.id);
    console.log(`   ✓ Resumed, state: ${info?.state}`);
  } catch (e) {
    console.log(`   ✗ Failed: ${e}`);
    return false;
  }

  // Test 11: Cancel SubAgent
  console.log('\n11. Testing cancelSubagent()...');
  try {
    orch.cancelSubagent(handle.id);
    await sleep(200);
    const info = orch.getSubagentInfo(handle.id);
    console.log(`   ✓ Cancelled, state: ${info?.state}`);
  } catch (e) {
    console.log(`   ✗ Failed: ${e}`);
    return false;
  }

  console.log('\n' + '='.repeat(50));
  console.log('✓ All API tests passed!');
  console.log('='.repeat(50));
  return true;
}

testOrchestratorAPIs()
  .then((success) => {
    process.exit(success ? 0 : 1);
  })
  .catch((error) => {
    console.error('Error:', error);
    process.exit(1);
  });

File: sdk/python/examples/test_apis.py
#!/usr/bin/env python3
"""
Test script to verify Orchestrator monitoring APIs are working correctly.
This test uses placeholder execution (no real LLM calls).
"""

import sys
import time


def test_orchestrator_apis():
    """Test all Orchestrator monitoring APIs"""
    print("Testing Orchestrator Monitoring APIs...\n")

    try:
        from a3s_code import Orchestrator, SubAgentConfig
        print("✓ Import successful")
    except ImportError as e:
        print(f"✗ Import failed: {e}")
        print("\nNote: This is expected if a3s-code is not installed.")
        print("The API definitions are correct, but the package needs to be built and installed.")
        return True

    # Test 1: Create Orchestrator
    print("\n1. Testing Orchestrator.create()...")
    try:
        orch = Orchestrator.create()
        print("   ✓ Orchestrator created")
    except Exception as e:
        print(f"   ✗ Failed: {e}")
        return False

    # Test 2: Create SubAgentConfig
    print("\n2. Testing SubAgentConfig...")
    try:
        config = SubAgentConfig(
            agent_type="test",
            description="Test SubAgent",
            prompt="Test prompt",
            permissive=True,
            max_steps=3,
        )
        print("   ✓ SubAgentConfig created")
    except Exception as e:
        print(f"   ✗ Failed: {e}")
        return False

    # Test 3: Spawn SubAgent
    print("\n3. Testing spawn_subagent()...")
    try:
        handle = orch.spawn_subagent(config)
        print(f"   ✓ SubAgent spawned: {handle.id}")
    except Exception as e:
        print(f"   ✗ Failed: {e}")
        return False

    # Test 4: Active count
    print("\n4. Testing active_count()...")
    try:
        count = orch.active_count()
        print(f"   ✓ Active count: {count}")
    except Exception as e:
        print(f"   ✗ Failed: {e}")
        return False

    # Test 5: List SubAgents
    print("\n5. Testing list_subagents()...")
    try:
        subagents = orch.list_subagents()
        print(f"   ✓ Found {len(subagents)} SubAgent(s)")
        if len(subagents) > 0:
            info = subagents[0]
            print(f"      - ID: {info.id}")
            print(f"      - Type: {info.agent_type}")
            print(f"      - State: {info.state}")
            print(f"      - Description: {info.description}")
            if info.current_activity:
                print(f"      - Activity: {info.current_activity.activity_type}")
    except Exception as e:
        print(f"   ✗ Failed: {e}")
        return False

    # Test 6: Get SubAgent info
    print("\n6. Testing get_subagent_info()...")
    try:
        info = orch.get_subagent_info(handle.id)
        if info:
            print(f"   ✓ Got info for {info.id}")
            print(f"      - State: {info.state}")
            print(f"      - Created: {info.created_at}")
            print(f"      - Updated: {info.updated_at}")
        else:
            print("   ✗ No info returned")
            return False
    except Exception as e:
        print(f"   ✗ Failed: {e}")
        return False

    # Test 7: Get active activities
    print("\n7. Testing get_active_activities()...")
    try:
        activities = orch.get_active_activities()
        print(f"   ✓ Found {len(activities)} active activit(ies)")
        for subagent_id, activity in activities:
            print(f"      - {subagent_id}: {activity.activity_type}")
    except Exception as e:
        print(f"   ✗ Failed: {e}")
        return False

    # Test 8: Get all states
    print("\n8. Testing get_all_states()...")
    try:
        states = orch.get_all_states()
        print(f"   ✓ Found {len(states)} state(s)")
        for subagent_id, state in states:
            print(f"      - {subagent_id}: {state}")
    except Exception as e:
        print(f"   ✗ Failed: {e}")
        return False

    # Test 9: Pause SubAgent
    print("\n9. Testing pause_subagent()...")
    try:
        orch.pause_subagent(handle.id)
        time.sleep(0.2)
        info = orch.get_subagent_info(handle.id)
        print(f"   ✓ Paused, state: {info.state}")
    except Exception as e:
        print(f"   ✗ Failed: {e}")
        return False

    # Test 10: Resume SubAgent
    print("\n10. Testing resume_subagent()...")
    try:
        orch.resume_subagent(handle.id)
        time.sleep(0.2)
        info = orch.get_subagent_info(handle.id)
        print(f"   ✓ Resumed, state: {info.state}")
    except Exception as e:
        print(f"   ✗ Failed: {e}")
        return False

    # Test 11: Cancel SubAgent
    print("\n11. Testing cancel_subagent()...")
    try:
        orch.cancel_subagent(handle.id)
        time.sleep(0.2)
        info = orch.get_subagent_info(handle.id)
        print(f"   ✓ Cancelled, state: {info.state}")
    except Exception as e:
        print(f"   ✗ Failed: {e}")
        return False

    print("\n" + "=" * 50)
    print("✓ All API tests passed!")
    print("=" * 50)
    return True


if __name__ == "__main__":
    success = test_orchestrator_apis()
    sys.exit(0 if success else 1)
